### ISSUE TITLE
Differentiate with regard a subset of the input

### ISSUE THREAD

**Issue by de-gozaru** [pytorch/pytorch#95132]
### 🚀 The feature, motivation and pitch

Hi,

When I want to differentiate with regard to the first element in a batch, I get an error, as shown in the following example:

```python
n, d = 10, 3
lin1 = nn.Linear(d, d)

# case 1
x = torch.randn(d).requires_grad_()
y = lin1(x)
vec = torch.ones(d)
gr = torch.autograd.grad(y, x, vec, retain_graph=True)[0]
print(gr.shape)  # returns: torch.Size([3])

# case 2
x = torch.randn(n, d).requires_grad_()
y = lin1(x)
vec = torch.ones(d)
gr = torch.autograd.grad(y[0], x[0], vec, retain_graph=True)[0]

# Raises Error:
# RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.
```
so taking x[i] removes it from the computation graph, I tried to clone it or to use .narrow but in vain!

y[0] only depends on x[0], so I don’t want to compute the gradient with regard to the full input!

Is there a way to do this efficiently?

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by ezyang** [comment]
Perhaps you want per-sample gradients? Functorch can do that.

**Comment by soulitzer** [comment]
Yeah https://pytorch.org/functorch/stable/notebooks/per_sample_grads.html is probably what you want. Make sure your function is able to able to accept a single sample as input. Closing for now, feel free to reopen or ask a question on the forums if you run into issues.
